{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Default Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_expression_path = './dados_proteomica_com_foldchange.xlsx'\n",
    "cancer_sheet = 'Cancer-Rat'\n",
    "preeclam_sheet = 'Preeclampsia-Rat'\n",
    "\n",
    "string_cancer_path = './string_cancer.csv'\n",
    "string_preeclam_path = './String_pre_eclampsia.csv'\n",
    "\n",
    "kegg_cancer_path = './KEGG_cancer.csv'\n",
    "kegg_preeclam_path = './KEGG_pre_eclampsia.xlsx'\n",
    "\n",
    "output_path = Path('./output')\n",
    "output_path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_exp_cols = ['Query_protein', 'Gene_name', 'FC']\n",
    "rename_cancer_exp_cols = {'Query_protein':'protein', 'Gene_name':'gene', 'FC':'fc'}\n",
    "cancer_expression_data = pd.read_excel(diff_expression_path, sheet_name=cancer_sheet, usecols=cancer_exp_cols).rename(columns=rename_cancer_exp_cols)\n",
    "cancer_expression_data['regulated'] = np.where(cancer_expression_data['fc'] > 0, 'up', 'down')\n",
    "\n",
    "preeclam_exp_cols = ['protein_firstname', 'Gene_name_correct', 'FC']\n",
    "rename_preeclam_exp_cols = {'protein_firstname':'protein', 'Gene_name_correct':'gene', 'FC':'fc'}\n",
    "preeclam_expression_data = pd.read_excel(diff_expression_path, sheet_name=preeclam_sheet, usecols=preeclam_exp_cols).rename(columns=rename_preeclam_exp_cols)\n",
    "preeclam_expression_data['regulated'] = np.where(preeclam_expression_data['fc'] > 0, 'up', 'down')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Cancer NAs:\n",
      " protein      0\n",
      "gene         0\n",
      "fc           1\n",
      "regulated    0\n",
      "dtype: int64\n",
      "\n",
      "## Pre-eclam NAs:\n",
      " protein       0\n",
      "gene         10\n",
      "fc            0\n",
      "regulated     0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Drop rows missing gene or fold change\n",
    "\n",
    "print('## Cancer NAs:\\n', cancer_expression_data.isna().sum())\n",
    "print()\n",
    "print('## Pre-eclam NAs:\\n', preeclam_expression_data.isna().sum())\n",
    "\n",
    "cancer_expression_data.dropna(how='any', subset=['gene', 'fc'], inplace=True)\n",
    "preeclam_expression_data.dropna(how='any', subset=['gene', 'fc'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tuba1c    2\n",
       "Name: gene, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc</th>\n",
       "      <th>regulated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q6AYZ1</td>\n",
       "      <td>Tuba1c</td>\n",
       "      <td>-0.440370</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>A0A0H2UHM7</td>\n",
       "      <td>Tuba1c</td>\n",
       "      <td>-0.355256</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        protein    gene        fc regulated\n",
       "0        Q6AYZ1  Tuba1c -0.440370      down\n",
       "160  A0A0H2UHM7  Tuba1c -0.355256      down"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Cancer - Drop duplicated genes (?) (keep the first occurrence)\n",
    "display(cancer_expression_data['gene'].value_counts()[cancer_expression_data['gene'].value_counts() > 1])\n",
    "display(cancer_expression_data[cancer_expression_data['gene'] == 'Tuba1c'])\n",
    "cancer_expression_data.drop_duplicates(subset='gene', keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gbe1    2\n",
       "Gpx3    2\n",
       "Name: gene, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc</th>\n",
       "      <th>regulated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>A0A0G2JTB2</td>\n",
       "      <td>Gbe1</td>\n",
       "      <td>0.533101</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>A0A096MJY6</td>\n",
       "      <td>Gbe1</td>\n",
       "      <td>0.537608</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>A0A0G2K531</td>\n",
       "      <td>Gpx3</td>\n",
       "      <td>-0.320739</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>A0A0G2K8W9</td>\n",
       "      <td>Gpx3</td>\n",
       "      <td>-0.290763</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        protein  gene        fc regulated\n",
       "208  A0A0G2JTB2  Gbe1  0.533101        up\n",
       "209  A0A096MJY6  Gbe1  0.537608        up\n",
       "317  A0A0G2K531  Gpx3 -0.320739      down\n",
       "318  A0A0G2K8W9  Gpx3 -0.290763      down"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Pre-eclampsia - Drop duplicated genes (?) (keep the first occurrence)\n",
    "display(preeclam_expression_data['gene'].value_counts()[preeclam_expression_data['gene'].value_counts() > 1])\n",
    "display(preeclam_expression_data[(preeclam_expression_data['gene'] == 'Gbe1') | (preeclam_expression_data['gene'] == 'Gpx3')])\n",
    "preeclam_expression_data.drop_duplicates(subset='gene', keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Merge differential expression data (cancer + pre-eclampsia)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_expression_data = cancer_expression_data.merge(preeclam_expression_data, on='gene', suffixes=['_cancer', '_pre_eclampsia'], how='outer')\n",
    "merged_expression_data['cancer'] = np.where(merged_expression_data['fc_cancer'].notna(), 1, 0)\n",
    "merged_expression_data['pre_eclampsia'] = np.where(merged_expression_data['fc_pre_eclampsia'].notna(), 1, 0)\n",
    "merged_expression_data['both'] = np.where((merged_expression_data['cancer'] == 1) & (merged_expression_data['pre_eclampsia'] == 1), 1, 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_cancer</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc_cancer</th>\n",
       "      <th>regulated_cancer</th>\n",
       "      <th>protein_pre_eclampsia</th>\n",
       "      <th>fc_pre_eclampsia</th>\n",
       "      <th>regulated_pre_eclampsia</th>\n",
       "      <th>cancer</th>\n",
       "      <th>pre_eclampsia</th>\n",
       "      <th>both</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q6AYZ1</td>\n",
       "      <td>Tuba1c</td>\n",
       "      <td>-0.440370</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q9EPH1</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>-0.778146</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P06238</td>\n",
       "      <td>A2m</td>\n",
       "      <td>-0.711127</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G3V9J6</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>-0.557322</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P68136</td>\n",
       "      <td>Acta1</td>\n",
       "      <td>-0.284675</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P62738</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>-0.316151</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>D3ZRN3</td>\n",
       "      <td>Actbl2</td>\n",
       "      <td>-0.322669</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P68035</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>-0.284966</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P63259</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>-0.308877</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>Actn1</td>\n",
       "      <td>0.240738</td>\n",
       "      <td>up</td>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>0.263762</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_cancer    gene  fc_cancer regulated_cancer protein_pre_eclampsia  \\\n",
       "0         Q6AYZ1  Tuba1c  -0.440370             down                   NaN   \n",
       "1         Q9EPH1    A1bg  -0.778146             down                   NaN   \n",
       "2         P06238     A2m  -0.711127             down                   NaN   \n",
       "3         G3V9J6  Abcb1b  -0.557322             down                   NaN   \n",
       "4         P68136   Acta1  -0.284675             down                   NaN   \n",
       "5         P62738   Acta2  -0.316151             down                   NaN   \n",
       "6         D3ZRN3  Actbl2  -0.322669             down                   NaN   \n",
       "7        P68035    Actc1  -0.284966             down                   NaN   \n",
       "8        P63259    Actg2  -0.308877             down                   NaN   \n",
       "9         Q9Z1P2   Actn1   0.240738               up                Q9Z1P2   \n",
       "\n",
       "   fc_pre_eclampsia regulated_pre_eclampsia  cancer  pre_eclampsia  both  \n",
       "0               NaN                     NaN       1              0     0  \n",
       "1               NaN                     NaN       1              0     0  \n",
       "2               NaN                     NaN       1              0     0  \n",
       "3               NaN                     NaN       1              0     0  \n",
       "4               NaN                     NaN       1              0     0  \n",
       "5               NaN                     NaN       1              0     0  \n",
       "6               NaN                     NaN       1              0     0  \n",
       "7               NaN                     NaN       1              0     0  \n",
       "8               NaN                     NaN       1              0     0  \n",
       "9          0.263762                      up       1              1     1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_expression_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_cancer</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc_cancer</th>\n",
       "      <th>regulated_cancer</th>\n",
       "      <th>protein_pre_eclampsia</th>\n",
       "      <th>fc_pre_eclampsia</th>\n",
       "      <th>regulated_pre_eclampsia</th>\n",
       "      <th>cancer</th>\n",
       "      <th>pre_eclampsia</th>\n",
       "      <th>both</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>Actn1</td>\n",
       "      <td>0.240738</td>\n",
       "      <td>up</td>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>0.263762</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Q6TUG0</td>\n",
       "      <td>Dnajb11</td>\n",
       "      <td>-0.428988</td>\n",
       "      <td>down</td>\n",
       "      <td>Q6TUG0</td>\n",
       "      <td>0.442867</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Q641Z6</td>\n",
       "      <td>Ehd1</td>\n",
       "      <td>-0.450791</td>\n",
       "      <td>down</td>\n",
       "      <td>Q641Z6</td>\n",
       "      <td>-0.294619</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Q8R3Z7</td>\n",
       "      <td>Ehd4</td>\n",
       "      <td>0.576265</td>\n",
       "      <td>up</td>\n",
       "      <td>Q8R3Z7</td>\n",
       "      <td>-0.293523</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>G3V843</td>\n",
       "      <td>F2</td>\n",
       "      <td>-0.553859</td>\n",
       "      <td>down</td>\n",
       "      <td>G3V843</td>\n",
       "      <td>0.293442</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>P20059</td>\n",
       "      <td>Hpx</td>\n",
       "      <td>0.444360</td>\n",
       "      <td>up</td>\n",
       "      <td>P20059</td>\n",
       "      <td>0.555551</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>P11762</td>\n",
       "      <td>Lgals1</td>\n",
       "      <td>-0.287505</td>\n",
       "      <td>down</td>\n",
       "      <td>P11762</td>\n",
       "      <td>0.315815</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Q9R063</td>\n",
       "      <td>Prdx5</td>\n",
       "      <td>-0.360268</td>\n",
       "      <td>down</td>\n",
       "      <td>A0A0G2JSS8</td>\n",
       "      <td>-0.353523</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>G3V8T5</td>\n",
       "      <td>Ruvbl2</td>\n",
       "      <td>-0.296228</td>\n",
       "      <td>down</td>\n",
       "      <td>G3V8T5</td>\n",
       "      <td>0.217608</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Q794F9</td>\n",
       "      <td>Slc3a2</td>\n",
       "      <td>-0.531778</td>\n",
       "      <td>down</td>\n",
       "      <td>Q794F9</td>\n",
       "      <td>-0.290488</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Q4QRB4</td>\n",
       "      <td>Tubb3</td>\n",
       "      <td>-0.148525</td>\n",
       "      <td>down</td>\n",
       "      <td>Q4QRB4</td>\n",
       "      <td>-0.325208</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>D4A8H3</td>\n",
       "      <td>Uba6</td>\n",
       "      <td>0.192763</td>\n",
       "      <td>up</td>\n",
       "      <td>D4A8H3</td>\n",
       "      <td>-0.346560</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_cancer     gene  fc_cancer regulated_cancer protein_pre_eclampsia  \\\n",
       "9           Q9Z1P2    Actn1   0.240738               up                Q9Z1P2   \n",
       "48          Q6TUG0  Dnajb11  -0.428988             down                Q6TUG0   \n",
       "51          Q641Z6     Ehd1  -0.450791             down                Q641Z6   \n",
       "52          Q8R3Z7     Ehd4   0.576265               up                Q8R3Z7   \n",
       "60          G3V843       F2  -0.553859             down                G3V843   \n",
       "76          P20059      Hpx   0.444360               up                P20059   \n",
       "89          P11762   Lgals1  -0.287505             down                P11762   \n",
       "117         Q9R063    Prdx5  -0.360268             down            A0A0G2JSS8   \n",
       "141         G3V8T5   Ruvbl2  -0.296228             down                G3V8T5   \n",
       "148         Q794F9   Slc3a2  -0.531778             down                Q794F9   \n",
       "162         Q4QRB4    Tubb3  -0.148525             down                Q4QRB4   \n",
       "168         D4A8H3     Uba6   0.192763               up                D4A8H3   \n",
       "\n",
       "     fc_pre_eclampsia regulated_pre_eclampsia  cancer  pre_eclampsia  both  \n",
       "9            0.263762                      up       1              1     1  \n",
       "48           0.442867                      up       1              1     1  \n",
       "51          -0.294619                    down       1              1     1  \n",
       "52          -0.293523                    down       1              1     1  \n",
       "60           0.293442                      up       1              1     1  \n",
       "76           0.555551                      up       1              1     1  \n",
       "89           0.315815                      up       1              1     1  \n",
       "117         -0.353523                    down       1              1     1  \n",
       "141          0.217608                      up       1              1     1  \n",
       "148         -0.290488                    down       1              1     1  \n",
       "162         -0.325208                    down       1              1     1  \n",
       "168         -0.346560                    down       1              1     1  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Verify common genes\n",
    "merged_expression_data[merged_expression_data['fc_cancer'].notna() & merged_expression_data['fc_pre_eclampsia'].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Cancer = 175, # Pre-eclampsia = 314, # Common = 12\n"
     ]
    }
   ],
   "source": [
    "# Count number of genes\n",
    "ct_cancer = sum(merged_expression_data['cancer'] == 1)\n",
    "ct_preeclam = sum(merged_expression_data['pre_eclampsia'] == 1)\n",
    "ct_both = sum(merged_expression_data['both'] == 1)\n",
    "print(f'# Cancer = {ct_cancer}, # Pre-eclampsia = {ct_preeclam}, # Common = {ct_both}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "exp_data_fn = 'expression_data'\n",
    "cancer_expression_data.to_csv(Path(output_path, f'{exp_data_fn}_cancer.csv'), index=False)\n",
    "preeclam_expression_data.to_csv(Path(output_path, f'{exp_data_fn}_pre_eclampsia.csv'), index=False)\n",
    "merged_expression_data.to_csv(Path(output_path, f'{exp_data_fn}_merged.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load String-db data (PPI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "string_cols = ['#node1', 'node2', 'node1_string_id', 'node2_string_id']\n",
    "rename_string_cols = {'#node1': 'node1', 'node1_string_id':'node1_id', 'node2_string_id':'node2_id'}\n",
    "\n",
    "# Note: current string files are separated by semicolon\n",
    "cancer_ppi_data = pd.read_csv(string_cancer_path, usecols=string_cols, sep=';').rename(columns=rename_string_cols)\n",
    "preeclam_ppi_data = pd.read_csv(string_preeclam_path, usecols=string_cols, sep=';').rename(columns=rename_string_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000050322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Tln1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000022401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Pxn</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000043928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000050322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Myl6</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000070200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Myh14</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000072636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Vcl</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000074748</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node1  node2                  node1_id                  node2_id\n",
       "0  Acta1  Actc1  10116.ENSRNOP00000024084  10116.ENSRNOP00000011773\n",
       "1  Acta1  Acta2  10116.ENSRNOP00000024084  10116.ENSRNOP00000073101\n",
       "2  Acta1  Actg2  10116.ENSRNOP00000024084  10116.ENSRNOP00000050322\n",
       "3  Acta2  Actc1  10116.ENSRNOP00000073101  10116.ENSRNOP00000011773\n",
       "4  Acta2   Tln1  10116.ENSRNOP00000073101  10116.ENSRNOP00000022401\n",
       "5  Acta2    Pxn  10116.ENSRNOP00000073101  10116.ENSRNOP00000043928\n",
       "6  Acta2  Actg2  10116.ENSRNOP00000073101  10116.ENSRNOP00000050322\n",
       "7  Acta2   Myl6  10116.ENSRNOP00000073101  10116.ENSRNOP00000070200\n",
       "8  Acta2  Myh14  10116.ENSRNOP00000073101  10116.ENSRNOP00000072636\n",
       "9  Acta2    Vcl  10116.ENSRNOP00000073101  10116.ENSRNOP00000074748"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cancer -> # Total genes = 175, # Genes in PPI = 104\n",
      "Genes NOT in ppi (71):\n",
      "{'Ubap2l', 'Dstn', 'Anxa2', 'Vat1', 'Alpp', 'Rpia', 'Rap2c', 'Tubb4b', 'Prmt8', 'Cyb5b', 'Afm', 'Bsg', 'Atic', 'Psap', 'Ddost', 'Lgals1', 'Cdv3', 'Mug2', 'Ndrg1', 'Emc2', 'Lonp1', 'Car1', 'Atxn10', 'Slc25a10', 'Cntrl', 'Ddx17', 'Hpx', 'Etfb', 'Rilpl1', 'Cpsf6', 'Nefl', 'Pfkl', 'Stk26', 'Serpina3n', 'Aldh2', 'Rab33b', 'Dpysl2', 'Csrp1', 'Rragb', 'Dbnl', 'Aimp1', 'Gstm4', 'Prdx4', 'Fkbp14', 'Uchl4', 'Abcb1b', 'Atp1b1', 'Hbb-b2', 'Mat2a', 'Hax1', 'Cand1', 'Hba', 'Rab43', 'Itih2', 'Spata1', 'A1bg', 'Api5', 'A2m', 'En2', 'Ak2', 'Ptbp1', 'Olfm4', 'Cotl1', 'Tgm2', 'Gnpda1', 'Sfxn3', 'Mug1', 'Rbp4', 'Slc3a2', 'Vapa', 'Krt18'}\n"
     ]
    }
   ],
   "source": [
    "display(cancer_ppi_data.head(10))\n",
    "\n",
    "cancer_genes_in_ppi = set(cancer_ppi_data['node1'].values.tolist() + cancer_ppi_data['node2'].values.tolist())\n",
    "cancer_all_genes = set(cancer_expression_data['gene'].values.tolist())\n",
    "print(f'Cancer -> # Total genes = {len(cancer_all_genes)}, # Genes in PPI = {len(cancer_genes_in_ppi)}')\n",
    "\n",
    "print(f'Genes NOT in ppi ({len(cancer_all_genes - cancer_genes_in_ppi)}):')\n",
    "print(cancer_all_genes - cancer_genes_in_ppi)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Itga6</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000002075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Mcam</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000010464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Efhd2</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000018864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Itga2b</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000052051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ada</td>\n",
       "      <td>Nt5c3a</td>\n",
       "      <td>10116.ENSRNOP00000014151</td>\n",
       "      <td>10116.ENSRNOP00000071357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ada</td>\n",
       "      <td>Adk</td>\n",
       "      <td>10116.ENSRNOP00000014151</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Ak2</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000000134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Jchain</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000004866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Upp1</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000006765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Nt5c3a</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000071357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node1   node2                  node1_id                  node2_id\n",
       "0  Actn1   Itga6  10116.ENSRNOP00000068851  10116.ENSRNOP00000002075\n",
       "1  Actn1    Mcam  10116.ENSRNOP00000068851  10116.ENSRNOP00000010464\n",
       "2  Actn1   Efhd2  10116.ENSRNOP00000068851  10116.ENSRNOP00000018864\n",
       "3  Actn1  Itga2b  10116.ENSRNOP00000068851  10116.ENSRNOP00000052051\n",
       "4    Ada  Nt5c3a  10116.ENSRNOP00000014151  10116.ENSRNOP00000071357\n",
       "5    Ada     Adk  10116.ENSRNOP00000014151  10116.ENSRNOP00000073983\n",
       "6    Adk     Ak2  10116.ENSRNOP00000073983  10116.ENSRNOP00000000134\n",
       "7    Adk  Jchain  10116.ENSRNOP00000073983  10116.ENSRNOP00000004866\n",
       "8    Adk    Upp1  10116.ENSRNOP00000073983  10116.ENSRNOP00000006765\n",
       "9    Adk  Nt5c3a  10116.ENSRNOP00000073983  10116.ENSRNOP00000071357"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cancer -> # Total genes = 314, # Genes in PPI = 153\n",
      "Genes NOT in ppi (200):\n",
      "{'Uggt1 ', 'Abcb4', 'Snx12', 'Kif2a ', 'Etfdh ', 'Slc40a1', 'LOC102557244', 'Pappa2', 'Cmas', 'Stmn1', 'Plvap', 'Mcam ', 'Matr3', 'Tspan9', 'Tfg ', 'Prl3d4', 'Krt19 ', 'Nt5c3a ', 'RGD1562844', 'Cox5b', 'Hsd17b11 ', 'Camk2d', 'Psg29 ', 'Uso1 ', 'Lad1', 'Man2a1', 'Tram2', 'Npc2', 'Eps8l2', 'Serpinb9', 'Pdxk', 'Tes', 'St6gal1 ', 'Tnfaip8', 'Pdlim4 ', 'Pltp', 'Fam122b', 'Pecam', 'RT1-CE1 ', 'Lman1 ', 'Grn', 'Ctla2a', 'Ccdc51', 'Lgalsl', 'Prxl2a ', 'Tbl2 ', 'C4bpa', 'Pgk1 ', 'Hyou1 ', 'Ppl', 'Slc2a1', 'Gab1', 'Atp2b1', 'Pon2', 'Emb ', 'Creld2', 'Aldh2 ', 'Myo18a', 'Phactr4', 'Lgals5', 'Nostrin', 'Psg19', 'Sbsn', 'Rps15 ', 'Crtap ', 'Pitrm1', 'LOC299282', 'Dpp3', 'Xdh', 'Ckap4', 'Prl6a1', 'Fblim1', 'Pcmt1', 'Nln', 'Hpx', 'Hmgb3 ', 'Nars1 ', 'Ak2 ', 'Prl4a1', 'Colec12 ', 'Csrp2', 'Sgpl1 ', 'Naxd ', 'Inf2', 'Cpne1', 'Sec11a ', 'Kng1', 'Mustn1', 'Lrrc59 ', 'Hp', 'Orm1', 'Timp3', 'Lima1 ', 'Agt ', 'Sptan1 ', 'Ces1c ', 'Upp1 ', 'AABR07060872.1', 'Lta4h', 'Arsb', 'Epb41l3 ', 'Glg1', 'Aqp1 ', 'Cotl1 ', 'Casp6', 'Prkar2a', 'Lipa', 'Arpp19', 'Btf3 ', 'Sparc', 'Gar1', 'Scamp1', 'Ero1a ', 'Cobll1', 'Retsat', 'Utrn ', 'Ctsql2', 'Ankrd13a ', 'Lamp1 ', 'Hdhd2 ', 'Mesd ', 'Plcd1', 'Prss8', 'Abhd12', 'Ralb ', 'Slc38a4 ', 'Dnajc3 ', 'Prrc1', 'Vnn1', 'Erp44', 'Hsp90b1 ', 'Ank3', 'Ngp', 'Cnn3', 'Efhd2 ', 'Septin5 ', 'Ctsm', 'Hspa5 ', 'Krt7', 'Fndc3a', 'Cavin1 ', 'Fbln1', 'Dbnl ', 'Syap1', 'Lrpap1', 'Itih1 ', 'Vcpip1 ', 'C4a', 'Rbp1 ', 'Pygl ', 'Gnb4', 'Abracl', 'Cd5l ', 'Fnta', 'Prl8a2', 'Ddah1 ', 'Adh1 ', 'Tfrc ', 'Fam136a', 'Aif1l', 'Slc4a2 ', 'Creg1', 'Ctsr', 'Lgals1', 'Ptpn23', 'Tapbp ', 'Hadh ', 'Slc27a3', 'Lman2 ', 'Scarb2 ', 'Cd55 ', 'Bgn', 'Lzic', 'Fkbp5 ', 'Dnajb1 ', 'Ceacam11 ', 'Ifnk ', 'Prl8a9 ', 'Eml2 ', 'Maged2', 'Tmpo ', 'Igfbp2 ', 'Klhl22', 'Loxl2', 'Ldhb ', 'Sh3bgrl', 'Uba6', 'Chdh', 'Manf ', 'Tm9sf2', 'Sars1 ', 'Tuba4a ', 'Itsn1 ', 'Limd1', 'Cubn', 'Psgb1', 'Hsph1 ', 'Cbr1l1', 'Afap1', 'Cand2'}\n"
     ]
    }
   ],
   "source": [
    "display(preeclam_ppi_data.head(10))\n",
    "\n",
    "preeclam_genes_in_ppi = set(preeclam_ppi_data['node1'].values.tolist() + preeclam_ppi_data['node2'].values.tolist())\n",
    "preeclam_all_genes = set(preeclam_expression_data['gene'].values.tolist())\n",
    "print(f'Cancer -> # Total genes = {len(preeclam_all_genes)}, # Genes in PPI = {len(preeclam_genes_in_ppi)}')\n",
    "\n",
    "print(f'Genes NOT in ppi ({len(preeclam_all_genes - preeclam_genes_in_ppi)}):')\n",
    "print(preeclam_all_genes - preeclam_genes_in_ppi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cancer -> # Total genes = 314, # Genes in PPI = 106\n",
      "Genes NOT in ppi (208):\n",
      "{'Uggt1 ', 'Abcb4', 'Snx12', 'Kif2a ', 'Etfdh ', 'Slc40a1', 'LOC102557244', 'Pappa2', 'Cmas', 'Stmn1', 'Plvap', 'Mcam ', 'Matr3', 'Tspan9', 'Tfg ', 'Prl3d4', 'Krt19 ', 'Nt5c3a ', 'RGD1562844', 'Cox5b', 'Hsd17b11 ', 'Camk2d', 'Psg29 ', 'Uso1 ', 'Lad1', 'Man2a1', 'Erp29', 'Tram2', 'Npc2', 'Eps8l2', 'Serpinb9', 'Pdxk', 'Tes', 'St6gal1 ', 'Tnfaip8', 'Pdlim4 ', 'Pltp', 'Fam122b', 'Pecam', 'RT1-CE1 ', 'Lman1 ', 'Grn', 'Ctla2a', 'Ccdc51', 'Lgalsl', 'Prxl2a ', 'Tbl2 ', 'C4bpa', 'Pgk1 ', 'Hyou1 ', 'Ppl', 'Slc2a1', 'Gab1', 'Atp2b1', 'Pon2', 'Emb ', 'Creld2', 'Aldh2 ', 'Myo18a', 'Phactr4', 'Lgals5', 'Nostrin', 'Psg19', 'Sbsn', 'Rps15 ', 'Crtap ', 'Pitrm1', 'LOC299282', 'Dpp3', 'Xdh', 'Ckap4', 'Prl6a1', 'Fblim1', 'Pcmt1', 'Nln', 'Hpx', 'Hmgb3 ', 'Nars1 ', 'Ak2 ', 'Prl4a1', 'Colec12 ', 'Csrp2', 'Sgpl1 ', 'Naxd ', 'Inf2', 'Cpne1', 'Sec11a ', 'Kng1', 'Mustn1', 'Lrrc59 ', 'Hp', 'Orm1', 'Timp3', 'Lima1 ', 'Agt ', 'Sptan1 ', 'Ces1c ', 'Upp1 ', 'AABR07060872.1', 'Lta4h', 'Arsb', 'Epb41l3 ', 'Glg1', 'Aqp1 ', 'Cotl1 ', 'Casp6', 'Prkar2a', 'Lipa', 'Arpp19', 'Btf3 ', 'Sparc', 'Gar1', 'Scamp1', 'Ero1a ', 'Tubb3', 'Cobll1', 'Retsat', 'Utrn ', 'Ctsql2', 'Ankrd13a ', 'Lamp1 ', 'Hdhd2 ', 'Mesd ', 'Plcd1', 'Ppid', 'Prss8', 'Abhd12', 'Ralb ', 'Igf2r', 'Slc38a4 ', 'Dnajc3 ', 'Prrc1', 'Vnn1', 'Erp44', 'Hsp90b1 ', 'Ank3', 'Ngp', 'Cnn3', 'Efhd2 ', 'Septin5 ', 'Ctsm', 'Hspa5 ', 'Krt7', 'Fndc3a', 'Cavin1 ', 'Fbln1', 'Dbnl ', 'Syap1', 'Lrpap1', 'Itih1 ', 'Vcpip1 ', 'C4a', 'Rbp1 ', 'Pygl ', 'Gnb4', 'Abracl', 'Cd5l ', 'Fnta', 'Prl8a2', 'Ddah1 ', 'Adh1 ', 'Tfrc ', 'Cgm4', 'Fam136a', 'Aif1l', 'Slc4a2 ', 'Creg1', 'Ctsr', 'Lgals1', 'Ptpn23', 'Tapbp ', 'Hadh ', 'Slc27a3', 'Lman2 ', 'Scarb2 ', 'Cd55 ', 'Bgn', 'Lzic', 'Fkbp5 ', 'Dnajb1 ', 'Ceacam11 ', 'Ifnk ', 'Prl8a9 ', 'Eml2 ', 'Maged2', 'Hspa12b', 'Tmpo ', 'Igfbp2 ', 'Klhl22', 'Loxl2', 'Ldhb ', 'Sh3bgrl', 'Uba6', 'Chdh', 'Manf ', 'Kif13b', 'Tm9sf2', 'Sars1 ', 'Tuba4a ', 'Itsn1 ', 'Limd1', 'Cubn', 'Psgb1', 'Hsph1 ', 'Cbr1l1', 'Stk10', 'Afap1', 'Cand2'}\n"
     ]
    }
   ],
   "source": [
    "###### Verificar como podem existir Genes no STRING que não estão na tabela de expressão diferencial\n",
    "preeclam_ppi_data = preeclam_ppi_data[preeclam_ppi_data['node1'].isin(preeclam_all_genes) & preeclam_ppi_data['node2'].isin(preeclam_all_genes)]\n",
    "\n",
    "\n",
    "\n",
    "preeclam_genes_in_ppi = set(preeclam_ppi_data['node1'].values.tolist() + preeclam_ppi_data['node2'].values.tolist())\n",
    "preeclam_all_genes = set(preeclam_expression_data['gene'].values.tolist())\n",
    "print(f'Cancer -> # Total genes = {len(preeclam_all_genes)}, # Genes in PPI = {len(preeclam_genes_in_ppi)}')\n",
    "\n",
    "print(f'Genes NOT in ppi ({len(preeclam_all_genes - preeclam_genes_in_ppi)}):')\n",
    "print(preeclam_all_genes - preeclam_genes_in_ppi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000050322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Tln1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000022401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Rps9</td>\n",
       "      <td>Tmed3</td>\n",
       "      <td>10116.ENSRNOP00000074779</td>\n",
       "      <td>10116.ENSRNOP00000018603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>Ruvbl2</td>\n",
       "      <td>Sdf2l1</td>\n",
       "      <td>10116.ENSRNOP00000028217</td>\n",
       "      <td>10116.ENSRNOP00000002540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Ruvbl2</td>\n",
       "      <td>Uchl5</td>\n",
       "      <td>10116.ENSRNOP00000028217</td>\n",
       "      <td>10116.ENSRNOP00000072164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>Sec61a1</td>\n",
       "      <td>Sec63</td>\n",
       "      <td>10116.ENSRNOP00000018455</td>\n",
       "      <td>10116.ENSRNOP00000069126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>Tmed3</td>\n",
       "      <td>Tmed7</td>\n",
       "      <td>10116.ENSRNOP00000018603</td>\n",
       "      <td>10116.ENSRNOP00000004942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>359 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       node1   node2                  node1_id                  node2_id\n",
       "0      Acta1   Actc1  10116.ENSRNOP00000024084  10116.ENSRNOP00000011773\n",
       "1      Acta1   Acta2  10116.ENSRNOP00000024084  10116.ENSRNOP00000073101\n",
       "2      Acta1   Actg2  10116.ENSRNOP00000024084  10116.ENSRNOP00000050322\n",
       "3      Acta2   Actc1  10116.ENSRNOP00000073101  10116.ENSRNOP00000011773\n",
       "4      Acta2    Tln1  10116.ENSRNOP00000073101  10116.ENSRNOP00000022401\n",
       "..       ...     ...                       ...                       ...\n",
       "354     Rps9   Tmed3  10116.ENSRNOP00000074779  10116.ENSRNOP00000018603\n",
       "355   Ruvbl2  Sdf2l1  10116.ENSRNOP00000028217  10116.ENSRNOP00000002540\n",
       "356   Ruvbl2   Uchl5  10116.ENSRNOP00000028217  10116.ENSRNOP00000072164\n",
       "357  Sec61a1   Sec63  10116.ENSRNOP00000018455  10116.ENSRNOP00000069126\n",
       "358    Tmed3   Tmed7  10116.ENSRNOP00000018603  10116.ENSRNOP00000004942\n",
       "\n",
       "[359 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "node1    node2 \n",
       "Dnajb11  Ruvbl2    2\n",
       "Ehd1     Ehd4      2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge STRING networks and check/remove for duplicates\n",
    "merged_ppi_data = pd.concat([cancer_ppi_data, preeclam_ppi_data], axis=0, join='outer', ignore_index=True)\n",
    "display(merged_ppi_data)\n",
    "print()\n",
    "merged_ppi_data.pivot_table(index = ['node1', 'node2'], aggfunc ='size')[merged_ppi_data.pivot_table(index = ['node1', 'node2'], aggfunc ='size') > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of Edges = 357, # of Nodes = 106\n"
     ]
    }
   ],
   "source": [
    "merged_ppi_data.drop_duplicates(keep='first', ignore_index=True, inplace=True)\n",
    "unique_nodes = set(preeclam_ppi_data['node1'].values.tolist() + preeclam_ppi_data['node2'].values.tolist())\n",
    "print(f'# of Edges = {len(merged_ppi_data)}, # of Nodes = {len(unique_nodes)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "ppi_fn = 'ppi_data'\n",
    "cancer_ppi_data.to_csv(Path(output_path, f'./{ppi_fn}_cancer.csv'), index=False)\n",
    "preeclam_ppi_data.to_csv(Path(output_path, f'./{ppi_fn}_pre_eclampsia.csv'), index=False)\n",
    "merged_ppi_data.to_csv(Path(output_path, f'./{ppi_fn}_merged.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load and process KEGG enrichment (pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_rename_cols = {\n",
    "    '#term ID' : 'pathway_id',\n",
    "    'term description' : 'pathway_name',\n",
    "    'observed gene count' : 'observed_gene_count',\n",
    "    'background gene count' : 'background_gene_count',\n",
    "    'strength' : 'strength',\n",
    "    'false discovery rate' : 'fdr',\n",
    "    'matching proteins in your network (labels)' : 'gene_labels_in'\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tuba4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Ctsl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>rno04144</td>\n",
       "      <td>Endocytosis</td>\n",
       "      <td>Ap2b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Pfkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Rpia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Mat2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Cs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                 pathway_name gene_labels_in\n",
       "0     rno04145                    Phagosome           Tfrc\n",
       "1     rno04145                    Phagosome         Tuba4a\n",
       "2     rno04145                    Phagosome          Tubb3\n",
       "3     rno04145                    Phagosome          Tubb6\n",
       "4     rno04145                    Phagosome           Ctsl\n",
       "..         ...                          ...            ...\n",
       "112   rno04144                  Endocytosis          Ap2b1\n",
       "113   rno01230  Biosynthesis of amino acids           Pfkl\n",
       "114   rno01230  Biosynthesis of amino acids           Rpia\n",
       "115   rno01230  Biosynthesis of amino acids          Mat2a\n",
       "116   rno01230  Biosynthesis of amino acids             Cs\n",
       "\n",
       "[117 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_pathway_data = pd.read_csv(kegg_cancer_path, sep=';', usecols=list(pathway_rename_cols.keys())).rename(columns=pathway_rename_cols)\n",
    "cancer_pathway_net = cancer_pathway_data[['pathway_id', 'pathway_name', 'gene_labels_in']].copy()\n",
    "cancer_pathway_net['gene_labels_in'] = cancer_pathway_net['gene_labels_in'].str.split(',')\n",
    "cancer_pathway_net = cancer_pathway_net.explode(column='gene_labels_in', ignore_index=True)\n",
    "cancer_pathway_net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Hsph1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Erp29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Dnajb11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Sec31a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Calr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>rno04922</td>\n",
       "      <td>Glucagon signaling pathway</td>\n",
       "      <td>Slc2a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc40a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Hmox1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc3a2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>189 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                                 pathway_name gene_labels_in\n",
       "0     rno04141  Protein processing in endoplasmic reticulum          Hsph1\n",
       "1     rno04141  Protein processing in endoplasmic reticulum          Erp29\n",
       "2     rno04141  Protein processing in endoplasmic reticulum        Dnajb11\n",
       "3     rno04141  Protein processing in endoplasmic reticulum         Sec31a\n",
       "4     rno04141  Protein processing in endoplasmic reticulum           Calr\n",
       "..         ...                                          ...            ...\n",
       "184   rno04922                   Glucagon signaling pathway         Slc2a1\n",
       "185   rno04216                                  Ferroptosis           Tfrc\n",
       "186   rno04216                                  Ferroptosis        Slc40a1\n",
       "187   rno04216                                  Ferroptosis          Hmox1\n",
       "188   rno04216                                  Ferroptosis         Slc3a2\n",
       "\n",
       "[189 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preeclam_pathway_data = pd.read_excel(kegg_preeclam_path, usecols=list(pathway_rename_cols.keys())).rename(columns=pathway_rename_cols)\n",
    "preeclam_pathway_net = preeclam_pathway_data[['pathway_id', 'pathway_name', 'gene_labels_in']].copy()\n",
    "preeclam_pathway_net['gene_labels_in'] = preeclam_pathway_net['gene_labels_in'].str.split(',')\n",
    "preeclam_pathway_net = preeclam_pathway_net.explode(column='gene_labels_in', ignore_index=True)\n",
    "preeclam_pathway_net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tuba4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Ctsl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>rno04922</td>\n",
       "      <td>Glucagon signaling pathway</td>\n",
       "      <td>Slc2a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc40a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Hmox1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc3a2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>306 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                pathway_name gene_labels_in\n",
       "0     rno04145                   Phagosome           Tfrc\n",
       "1     rno04145                   Phagosome         Tuba4a\n",
       "2     rno04145                   Phagosome          Tubb3\n",
       "3     rno04145                   Phagosome          Tubb6\n",
       "4     rno04145                   Phagosome           Ctsl\n",
       "..         ...                         ...            ...\n",
       "301   rno04922  Glucagon signaling pathway         Slc2a1\n",
       "302   rno04216                 Ferroptosis           Tfrc\n",
       "303   rno04216                 Ferroptosis        Slc40a1\n",
       "304   rno04216                 Ferroptosis          Hmox1\n",
       "305   rno04216                 Ferroptosis         Slc3a2\n",
       "\n",
       "[306 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "pathway_name                                 gene_labels_in\n",
       "Phagosome                                    Tfrc              2\n",
       "                                             Tuba4a            2\n",
       "                                             Tubb3             2\n",
       "Protein processing in endoplasmic reticulum  Dnajb11           2\n",
       "                                             Ero1a             2\n",
       "                                             Lman2             2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge both pathway nets and check/remove for duplicates\n",
    "\n",
    "merged_pathway_net = pd.concat([cancer_pathway_net, preeclam_pathway_net], axis=0, join='outer', ignore_index=True)\n",
    "display(merged_pathway_net)\n",
    "print()\n",
    "merged_pathway_net.pivot_table(index = ['pathway_name', 'gene_labels_in'], aggfunc ='size')[merged_pathway_net.pivot_table(index = ['pathway_name', 'gene_labels_in'], aggfunc ='size') > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_pathway_net.drop_duplicates(keep='first', ignore_index=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_net_fn = 'kegg_net'\n",
    "cancer_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_cancer.csv'), index=False)\n",
    "preeclam_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_pre_eclampsia.csv'), index=False)\n",
    "merged_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_merged.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
