{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Default Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_expression_path = Path('./dados_proteomica_com_foldchange.xlsx')\n",
    "cancer_sheet = 'Cancer-Rat'\n",
    "preeclam_sheet = 'Preeclampsia-Rat'\n",
    "\n",
    "string_cancer_path = Path('./String_cancer.csv')\n",
    "string_preeclam_path = Path('./String_pre_eclampsia.csv')\n",
    "\n",
    "kegg_cancer_path = Path('./KEGG_cancer.csv')\n",
    "kegg_preeclam_path = Path('./KEGG_pre_eclampsia.xlsx')\n",
    "kegg_common_path = Path('./KEGG_compartilhadas.xlsx')\n",
    "\n",
    "output_path = Path('./output')\n",
    "output_path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_exp_cols = ['Query_protein', 'Gene_name', 'FC']\n",
    "rename_cancer_exp_cols = {'Query_protein':'protein', 'Gene_name':'gene', 'FC':'fc'}\n",
    "cancer_expression_data = pd.read_excel(diff_expression_path, sheet_name=cancer_sheet, usecols=cancer_exp_cols).rename(columns=rename_cancer_exp_cols)\n",
    "cancer_expression_data['gene'] = cancer_expression_data['gene'].str.strip()\n",
    "cancer_expression_data['regulated'] = np.where(cancer_expression_data['fc'] > 0, 'up', 'down')\n",
    "\n",
    "\n",
    "preeclam_exp_cols = ['protein_firstname', 'Gene_name_correct', 'FC']\n",
    "rename_preeclam_exp_cols = {'protein_firstname':'protein', 'Gene_name_correct':'gene', 'FC':'fc'}\n",
    "preeclam_expression_data = pd.read_excel(diff_expression_path, sheet_name=preeclam_sheet, usecols=preeclam_exp_cols).rename(columns=rename_preeclam_exp_cols)\n",
    "preeclam_expression_data['gene'] = preeclam_expression_data['gene'].str.strip()\n",
    "preeclam_expression_data['regulated'] = np.where(preeclam_expression_data['fc'] > 0, 'up', 'down')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "177"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cancer_expression_data['gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "326"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(preeclam_expression_data['gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Cancer NAs:\n",
      " protein      0\n",
      "gene         0\n",
      "fc           1\n",
      "regulated    0\n",
      "dtype: int64\n",
      "\n",
      "## Pre-eclam NAs:\n",
      " protein       0\n",
      "gene         10\n",
      "fc            0\n",
      "regulated     0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Drop rows missing gene or fold change\n",
    "\n",
    "print('## Cancer NAs:\\n', cancer_expression_data.isna().sum())\n",
    "print()\n",
    "print('## Pre-eclam NAs:\\n', preeclam_expression_data.isna().sum())\n",
    "\n",
    "cancer_expression_data.dropna(how='any', subset=['gene', 'fc'], inplace=True)\n",
    "preeclam_expression_data.dropna(how='any', subset=['gene', 'fc'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tuba1c    2\n",
       "Name: gene, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc</th>\n",
       "      <th>regulated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q6AYZ1</td>\n",
       "      <td>Tuba1c</td>\n",
       "      <td>-0.440370</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>A0A0H2UHM7</td>\n",
       "      <td>Tuba1c</td>\n",
       "      <td>-0.355256</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        protein    gene        fc regulated\n",
       "0        Q6AYZ1  Tuba1c -0.440370      down\n",
       "160  A0A0H2UHM7  Tuba1c -0.355256      down"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Cancer - Drop duplicated genes (?) (keep the first occurrence)\n",
    "display(cancer_expression_data['gene'].value_counts()[cancer_expression_data['gene'].value_counts() > 1])\n",
    "display(cancer_expression_data[cancer_expression_data['gene'] == 'Tuba1c'])\n",
    "cancer_expression_data.drop_duplicates(subset='gene', keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gbe1    2\n",
       "Gpx3    2\n",
       "Name: gene, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc</th>\n",
       "      <th>regulated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>A0A0G2JTB2</td>\n",
       "      <td>Gbe1</td>\n",
       "      <td>0.533101</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>A0A096MJY6</td>\n",
       "      <td>Gbe1</td>\n",
       "      <td>0.537608</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>A0A0G2K531</td>\n",
       "      <td>Gpx3</td>\n",
       "      <td>-0.320739</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>A0A0G2K8W9</td>\n",
       "      <td>Gpx3</td>\n",
       "      <td>-0.290763</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        protein  gene        fc regulated\n",
       "208  A0A0G2JTB2  Gbe1  0.533101        up\n",
       "209  A0A096MJY6  Gbe1  0.537608        up\n",
       "317  A0A0G2K531  Gpx3 -0.320739      down\n",
       "318  A0A0G2K8W9  Gpx3 -0.290763      down"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Pre-eclampsia - Drop duplicated genes (?) (keep the first occurrence)\n",
    "display(preeclam_expression_data['gene'].value_counts()[preeclam_expression_data['gene'].value_counts() > 1])\n",
    "display(preeclam_expression_data[(preeclam_expression_data['gene'] == 'Gbe1') | (preeclam_expression_data['gene'] == 'Gpx3')])\n",
    "preeclam_expression_data.drop_duplicates(subset='gene', keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc</th>\n",
       "      <th>regulated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>F1LTN6</td>\n",
       "      <td>AABR07060872.1</td>\n",
       "      <td>0.754191</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein            gene        fc regulated\n",
       "315  F1LTN6  AABR07060872.1  0.754191        up"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preeclam_expression_data[preeclam_expression_data['gene'].str.contains(\".\", regex=False)]\n",
    "# preeclam_expression_data[preeclam_expression_data['gene'] == '.']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Merge differential expression data (cancer + pre-eclampsia)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_expression_data = cancer_expression_data.merge(preeclam_expression_data, on='gene', suffixes=['_cancer', '_pre_eclampsia'], how='outer')\n",
    "merged_expression_data['cancer'] = np.where(merged_expression_data['fc_cancer'].notna(), 1, 0)\n",
    "merged_expression_data['pre_eclampsia'] = np.where(merged_expression_data['fc_pre_eclampsia'].notna(), 1, 0)\n",
    "\n",
    "merged_expression_data['both'] = np.where((merged_expression_data['cancer'] == 1) & (merged_expression_data['pre_eclampsia'] == 1), 1, 0)\n",
    "\n",
    "merged_expression_data['present_in'] = np.where((merged_expression_data['cancer'] == 1) & (merged_expression_data['pre_eclampsia'] == 1), 1, 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_expression_data['present_in'] = np.where(merged_expression_data['cancer'] == 1, np.where(merged_expression_data['pre_eclampsia'] == 1, 'both', 'cancer'), 'pre_eclampsia')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_cancer</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc_cancer</th>\n",
       "      <th>regulated_cancer</th>\n",
       "      <th>protein_pre_eclampsia</th>\n",
       "      <th>fc_pre_eclampsia</th>\n",
       "      <th>regulated_pre_eclampsia</th>\n",
       "      <th>cancer</th>\n",
       "      <th>pre_eclampsia</th>\n",
       "      <th>both</th>\n",
       "      <th>present_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q6AYZ1</td>\n",
       "      <td>Tuba1c</td>\n",
       "      <td>-0.440370</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q9EPH1</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>-0.778146</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P06238</td>\n",
       "      <td>A2m</td>\n",
       "      <td>-0.711127</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G3V9J6</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>-0.557322</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P68136</td>\n",
       "      <td>Acta1</td>\n",
       "      <td>-0.284675</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P62738</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>-0.316151</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>D3ZRN3</td>\n",
       "      <td>Actbl2</td>\n",
       "      <td>-0.322669</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P68035</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>-0.284966</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P63259</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>-0.308877</td>\n",
       "      <td>down</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>Actn1</td>\n",
       "      <td>0.240738</td>\n",
       "      <td>up</td>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>0.263762</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_cancer    gene  fc_cancer regulated_cancer protein_pre_eclampsia  \\\n",
       "0         Q6AYZ1  Tuba1c  -0.440370             down                   NaN   \n",
       "1         Q9EPH1    A1bg  -0.778146             down                   NaN   \n",
       "2         P06238     A2m  -0.711127             down                   NaN   \n",
       "3         G3V9J6  Abcb1b  -0.557322             down                   NaN   \n",
       "4         P68136   Acta1  -0.284675             down                   NaN   \n",
       "5         P62738   Acta2  -0.316151             down                   NaN   \n",
       "6         D3ZRN3  Actbl2  -0.322669             down                   NaN   \n",
       "7        P68035    Actc1  -0.284966             down                   NaN   \n",
       "8        P63259    Actg2  -0.308877             down                   NaN   \n",
       "9         Q9Z1P2   Actn1   0.240738               up                Q9Z1P2   \n",
       "\n",
       "   fc_pre_eclampsia regulated_pre_eclampsia  cancer  pre_eclampsia  both  \\\n",
       "0               NaN                     NaN       1              0     0   \n",
       "1               NaN                     NaN       1              0     0   \n",
       "2               NaN                     NaN       1              0     0   \n",
       "3               NaN                     NaN       1              0     0   \n",
       "4               NaN                     NaN       1              0     0   \n",
       "5               NaN                     NaN       1              0     0   \n",
       "6               NaN                     NaN       1              0     0   \n",
       "7               NaN                     NaN       1              0     0   \n",
       "8               NaN                     NaN       1              0     0   \n",
       "9          0.263762                      up       1              1     1   \n",
       "\n",
       "  present_in  \n",
       "0     cancer  \n",
       "1     cancer  \n",
       "2     cancer  \n",
       "3     cancer  \n",
       "4     cancer  \n",
       "5     cancer  \n",
       "6     cancer  \n",
       "7     cancer  \n",
       "8     cancer  \n",
       "9       both  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_expression_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_cancer</th>\n",
       "      <th>gene</th>\n",
       "      <th>fc_cancer</th>\n",
       "      <th>regulated_cancer</th>\n",
       "      <th>protein_pre_eclampsia</th>\n",
       "      <th>fc_pre_eclampsia</th>\n",
       "      <th>regulated_pre_eclampsia</th>\n",
       "      <th>cancer</th>\n",
       "      <th>pre_eclampsia</th>\n",
       "      <th>both</th>\n",
       "      <th>present_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>Actn1</td>\n",
       "      <td>0.240738</td>\n",
       "      <td>up</td>\n",
       "      <td>Q9Z1P2</td>\n",
       "      <td>0.263762</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P29410</td>\n",
       "      <td>Ak2</td>\n",
       "      <td>-0.890588</td>\n",
       "      <td>down</td>\n",
       "      <td>A0A0G2JSG6</td>\n",
       "      <td>0.287707</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P11884</td>\n",
       "      <td>Aldh2</td>\n",
       "      <td>-0.461168</td>\n",
       "      <td>down</td>\n",
       "      <td>F1LN88</td>\n",
       "      <td>-0.441715</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>B0BNA5</td>\n",
       "      <td>Cotl1</td>\n",
       "      <td>0.821756</td>\n",
       "      <td>up</td>\n",
       "      <td>B0BNA5</td>\n",
       "      <td>-1.921076</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Q9JHL4</td>\n",
       "      <td>Dbnl</td>\n",
       "      <td>-1.110252</td>\n",
       "      <td>down</td>\n",
       "      <td>Q9JHL4</td>\n",
       "      <td>-0.311186</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Q6TUG0</td>\n",
       "      <td>Dnajb11</td>\n",
       "      <td>-0.428988</td>\n",
       "      <td>down</td>\n",
       "      <td>Q6TUG0</td>\n",
       "      <td>0.442867</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Q641Z6</td>\n",
       "      <td>Ehd1</td>\n",
       "      <td>-0.450791</td>\n",
       "      <td>down</td>\n",
       "      <td>Q641Z6</td>\n",
       "      <td>-0.294619</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Q8R3Z7</td>\n",
       "      <td>Ehd4</td>\n",
       "      <td>0.576265</td>\n",
       "      <td>up</td>\n",
       "      <td>Q8R3Z7</td>\n",
       "      <td>-0.293523</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Q8R4A1</td>\n",
       "      <td>Ero1a</td>\n",
       "      <td>0.306258</td>\n",
       "      <td>up</td>\n",
       "      <td>Q8R4A1</td>\n",
       "      <td>0.773918</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>G3V843</td>\n",
       "      <td>F2</td>\n",
       "      <td>-0.553859</td>\n",
       "      <td>down</td>\n",
       "      <td>G3V843</td>\n",
       "      <td>0.293442</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>P20059</td>\n",
       "      <td>Hpx</td>\n",
       "      <td>0.444360</td>\n",
       "      <td>up</td>\n",
       "      <td>P20059</td>\n",
       "      <td>0.555551</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>P42123</td>\n",
       "      <td>Ldhb</td>\n",
       "      <td>-0.263854</td>\n",
       "      <td>down</td>\n",
       "      <td>P42123</td>\n",
       "      <td>-0.494078</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>P11762</td>\n",
       "      <td>Lgals1</td>\n",
       "      <td>-0.287505</td>\n",
       "      <td>down</td>\n",
       "      <td>P11762</td>\n",
       "      <td>0.315815</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>B0BNG3</td>\n",
       "      <td>Lman2</td>\n",
       "      <td>-0.544716</td>\n",
       "      <td>down</td>\n",
       "      <td>B0BNG3</td>\n",
       "      <td>0.274321</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Q9R063</td>\n",
       "      <td>Prdx5</td>\n",
       "      <td>-0.360268</td>\n",
       "      <td>down</td>\n",
       "      <td>A0A0G2JSS8</td>\n",
       "      <td>-0.353523</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>G3V8T5</td>\n",
       "      <td>Ruvbl2</td>\n",
       "      <td>-0.296228</td>\n",
       "      <td>down</td>\n",
       "      <td>G3V8T5</td>\n",
       "      <td>0.217608</td>\n",
       "      <td>up</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Q794F9</td>\n",
       "      <td>Slc3a2</td>\n",
       "      <td>-0.531778</td>\n",
       "      <td>down</td>\n",
       "      <td>Q794F9</td>\n",
       "      <td>-0.290488</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>P16086</td>\n",
       "      <td>Sptan1</td>\n",
       "      <td>-0.416660</td>\n",
       "      <td>down</td>\n",
       "      <td>Q6IRK8</td>\n",
       "      <td>-0.286952</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Q99376</td>\n",
       "      <td>Tfrc</td>\n",
       "      <td>-0.516989</td>\n",
       "      <td>down</td>\n",
       "      <td>Q99376</td>\n",
       "      <td>-0.725451</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Q5XIF6</td>\n",
       "      <td>Tuba4a</td>\n",
       "      <td>-0.399533</td>\n",
       "      <td>down</td>\n",
       "      <td>Q5XIF6</td>\n",
       "      <td>-0.590092</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Q4QRB4</td>\n",
       "      <td>Tubb3</td>\n",
       "      <td>-0.148525</td>\n",
       "      <td>down</td>\n",
       "      <td>Q4QRB4</td>\n",
       "      <td>-0.325208</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>D4A8H3</td>\n",
       "      <td>Uba6</td>\n",
       "      <td>0.192763</td>\n",
       "      <td>up</td>\n",
       "      <td>D4A8H3</td>\n",
       "      <td>-0.346560</td>\n",
       "      <td>down</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_cancer     gene  fc_cancer regulated_cancer protein_pre_eclampsia  \\\n",
       "9           Q9Z1P2    Actn1   0.240738               up                Q9Z1P2   \n",
       "13          P29410      Ak2  -0.890588             down            A0A0G2JSG6   \n",
       "15          P11884    Aldh2  -0.461168             down                F1LN88   \n",
       "36          B0BNA5    Cotl1   0.821756               up                B0BNA5   \n",
       "44          Q9JHL4     Dbnl  -1.110252             down                Q9JHL4   \n",
       "48          Q6TUG0  Dnajb11  -0.428988             down                Q6TUG0   \n",
       "51          Q641Z6     Ehd1  -0.450791             down                Q641Z6   \n",
       "52          Q8R3Z7     Ehd4   0.576265               up                Q8R3Z7   \n",
       "58         Q8R4A1     Ero1a   0.306258               up                Q8R4A1   \n",
       "60          G3V843       F2  -0.553859             down                G3V843   \n",
       "76          P20059      Hpx   0.444360               up                P20059   \n",
       "88          P42123     Ldhb  -0.263854             down                P42123   \n",
       "89          P11762   Lgals1  -0.287505             down                P11762   \n",
       "90          B0BNG3    Lman2  -0.544716             down                B0BNG3   \n",
       "117         Q9R063    Prdx5  -0.360268             down            A0A0G2JSS8   \n",
       "141         G3V8T5   Ruvbl2  -0.296228             down                G3V8T5   \n",
       "148         Q794F9   Slc3a2  -0.531778             down                Q794F9   \n",
       "151         P16086   Sptan1  -0.416660             down                Q6IRK8   \n",
       "155         Q99376     Tfrc  -0.516989             down                Q99376   \n",
       "160         Q5XIF6   Tuba4a  -0.399533             down                Q5XIF6   \n",
       "162         Q4QRB4    Tubb3  -0.148525             down                Q4QRB4   \n",
       "168         D4A8H3     Uba6   0.192763               up                D4A8H3   \n",
       "\n",
       "     fc_pre_eclampsia regulated_pre_eclampsia  cancer  pre_eclampsia  both  \\\n",
       "9            0.263762                      up       1              1     1   \n",
       "13           0.287707                      up       1              1     1   \n",
       "15          -0.441715                    down       1              1     1   \n",
       "36          -1.921076                    down       1              1     1   \n",
       "44          -0.311186                    down       1              1     1   \n",
       "48           0.442867                      up       1              1     1   \n",
       "51          -0.294619                    down       1              1     1   \n",
       "52          -0.293523                    down       1              1     1   \n",
       "58           0.773918                      up       1              1     1   \n",
       "60           0.293442                      up       1              1     1   \n",
       "76           0.555551                      up       1              1     1   \n",
       "88          -0.494078                    down       1              1     1   \n",
       "89           0.315815                      up       1              1     1   \n",
       "90           0.274321                      up       1              1     1   \n",
       "117         -0.353523                    down       1              1     1   \n",
       "141          0.217608                      up       1              1     1   \n",
       "148         -0.290488                    down       1              1     1   \n",
       "151         -0.286952                    down       1              1     1   \n",
       "155         -0.725451                    down       1              1     1   \n",
       "160         -0.590092                    down       1              1     1   \n",
       "162         -0.325208                    down       1              1     1   \n",
       "168         -0.346560                    down       1              1     1   \n",
       "\n",
       "    present_in  \n",
       "9         both  \n",
       "13        both  \n",
       "15        both  \n",
       "36        both  \n",
       "44        both  \n",
       "48        both  \n",
       "51        both  \n",
       "52        both  \n",
       "58        both  \n",
       "60        both  \n",
       "76        both  \n",
       "88        both  \n",
       "89        both  \n",
       "90        both  \n",
       "117       both  \n",
       "141       both  \n",
       "148       both  \n",
       "151       both  \n",
       "155       both  \n",
       "160       both  \n",
       "162       both  \n",
       "168       both  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Verify common genes\n",
    "merged_expression_data[merged_expression_data['fc_cancer'].notna() & merged_expression_data['fc_pre_eclampsia'].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Cancer = 175, # Pre-eclampsia = 314, # Common = 22\n"
     ]
    }
   ],
   "source": [
    "# Count number of genes\n",
    "ct_cancer = sum(merged_expression_data['cancer'] == 1)\n",
    "ct_preeclam = sum(merged_expression_data['pre_eclampsia'] == 1)\n",
    "ct_both = sum(merged_expression_data['both'] == 1)\n",
    "print(f'# Cancer = {ct_cancer}, # Pre-eclampsia = {ct_preeclam}, # Common = {ct_both}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "exp_data_fn = 'expression_data'\n",
    "cancer_expression_data.to_csv(Path(output_path, f'{exp_data_fn}_cancer.csv'), index=False)\n",
    "preeclam_expression_data.to_csv(Path(output_path, f'{exp_data_fn}_pre_eclampsia.csv'), index=False)\n",
    "merged_expression_data.to_csv(Path(output_path, f'{exp_data_fn}_merged.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load String-db data (PPI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "string_cols = ['#node1', 'node2', 'node1_string_id', 'node2_string_id']\n",
    "rename_string_cols = {'#node1': 'node1', 'node1_string_id':'node1_id', 'node2_string_id':'node2_id'}\n",
    "\n",
    "# Note: current string files are separated by semicolon\n",
    "cancer_ppi_data = pd.read_csv(string_cancer_path, usecols=string_cols, sep=';').rename(columns=rename_string_cols)\n",
    "preeclam_ppi_data = pd.read_csv(string_preeclam_path, usecols=string_cols, sep=';').rename(columns=rename_string_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000050322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Tln1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000022401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Pxn</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000043928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000050322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Myl6</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000070200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Myh14</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000072636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Vcl</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000074748</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node1  node2                  node1_id                  node2_id\n",
       "0  Acta1  Actc1  10116.ENSRNOP00000024084  10116.ENSRNOP00000011773\n",
       "1  Acta1  Acta2  10116.ENSRNOP00000024084  10116.ENSRNOP00000073101\n",
       "2  Acta1  Actg2  10116.ENSRNOP00000024084  10116.ENSRNOP00000050322\n",
       "3  Acta2  Actc1  10116.ENSRNOP00000073101  10116.ENSRNOP00000011773\n",
       "4  Acta2   Tln1  10116.ENSRNOP00000073101  10116.ENSRNOP00000022401\n",
       "5  Acta2    Pxn  10116.ENSRNOP00000073101  10116.ENSRNOP00000043928\n",
       "6  Acta2  Actg2  10116.ENSRNOP00000073101  10116.ENSRNOP00000050322\n",
       "7  Acta2   Myl6  10116.ENSRNOP00000073101  10116.ENSRNOP00000070200\n",
       "8  Acta2  Myh14  10116.ENSRNOP00000073101  10116.ENSRNOP00000072636\n",
       "9  Acta2    Vcl  10116.ENSRNOP00000073101  10116.ENSRNOP00000074748"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cancer -> # Total genes = 176, # Genes in PPI = 104\n",
      "Genes NOT in ppi (72):\n",
      "{'Vapa', 'Pfkl', 'Tgm2', 'Ptbp1', 'Atp1b1', 'Ndrg1', 'Hbb-b2', 'Itih2', 'Api5', 'Lgals1', 'Rilpl1', 'Car1', 'Alpp', 'Aimp1', 'A2m', 'Rpia', 'Cyb5b', 'Prdx4', 'Tubb4b', 'Hba', 'Rragb', 'Cdv3', 'A1bg', 'Atxn10', 'Lonp1', 'Gnpda1', 'Ddx17', 'Krt18', 'Prmt8', 'Rab43', 'Nefl', 'En2', 'Cntrl', 'Mat2a', 'Gstm4', 'Hax1', 'Bsg', 'Atic', 'Dbnl', 'Aldh2', 'Mug2', 'Sfxn3', 'Abcb1b', 'Anxa2', 'Vat1', 'Uchl4', 'Psap', 'Ak2', 'Rab33b', 'Spata1', 'Stk26', 'Csrp1', 'Etfb', 'Slc25a10', 'Ddost', 'Rap2c', 'Afm', 'Cotl1', 'Emc2', 'Eno2', 'Ubap2l', 'Dpysl2', 'Hpx', 'Cand1', 'Slc3a2', 'Olfm4', 'Dstn', 'Rbp4', 'Mug1', 'Fkbp14', 'Cpsf6', 'Serpina3n'}\n"
     ]
    }
   ],
   "source": [
    "# Cancer\n",
    "display(cancer_ppi_data.head(10))\n",
    "\n",
    "cancer_genes_in_ppi = set(cancer_ppi_data['node1'].values.tolist() + cancer_ppi_data['node2'].values.tolist())\n",
    "cancer_all_genes = set(cancer_expression_data['gene'].values.tolist())\n",
    "\n",
    "print(f'Cancer -> # Total genes = {len(cancer_all_genes)}, # Genes in PPI = {len(cancer_genes_in_ppi)}')\n",
    "print(f'Genes NOT in ppi ({len(cancer_all_genes - cancer_genes_in_ppi)}):')\n",
    "print(cancer_all_genes - cancer_genes_in_ppi)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cancer -> # of Edges = 185, # of Nodes = 104\n"
     ]
    }
   ],
   "source": [
    "unique_nodes = set(cancer_ppi_data['node1'].values.tolist() + cancer_ppi_data['node2'].values.tolist())\n",
    "print(f'Cancer -> # of Edges = {len(cancer_ppi_data)}, # of Nodes = {len(unique_nodes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Itga6</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000002075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Mcam</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000010464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Efhd2</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000018864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Itga2b</td>\n",
       "      <td>10116.ENSRNOP00000068851</td>\n",
       "      <td>10116.ENSRNOP00000052051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ada</td>\n",
       "      <td>Nt5c3a</td>\n",
       "      <td>10116.ENSRNOP00000014151</td>\n",
       "      <td>10116.ENSRNOP00000071357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ada</td>\n",
       "      <td>Adk</td>\n",
       "      <td>10116.ENSRNOP00000014151</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Ak2</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000000134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Jchain</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000004866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Upp1</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000006765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Adk</td>\n",
       "      <td>Nt5c3a</td>\n",
       "      <td>10116.ENSRNOP00000073983</td>\n",
       "      <td>10116.ENSRNOP00000071357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node1   node2                  node1_id                  node2_id\n",
       "0  Actn1   Itga6  10116.ENSRNOP00000068851  10116.ENSRNOP00000002075\n",
       "1  Actn1    Mcam  10116.ENSRNOP00000068851  10116.ENSRNOP00000010464\n",
       "2  Actn1   Efhd2  10116.ENSRNOP00000068851  10116.ENSRNOP00000018864\n",
       "3  Actn1  Itga2b  10116.ENSRNOP00000068851  10116.ENSRNOP00000052051\n",
       "4    Ada  Nt5c3a  10116.ENSRNOP00000014151  10116.ENSRNOP00000071357\n",
       "5    Ada     Adk  10116.ENSRNOP00000014151  10116.ENSRNOP00000073983\n",
       "6    Adk     Ak2  10116.ENSRNOP00000073983  10116.ENSRNOP00000000134\n",
       "7    Adk  Jchain  10116.ENSRNOP00000073983  10116.ENSRNOP00000004866\n",
       "8    Adk    Upp1  10116.ENSRNOP00000073983  10116.ENSRNOP00000006765\n",
       "9    Adk  Nt5c3a  10116.ENSRNOP00000073983  10116.ENSRNOP00000071357"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pre-eclampsia -> # Total genes = 315, # Genes in PPI = 153\n",
      "Genes NOT in ppi (162):\n",
      "{nan, 'Ankrd13a', 'Tm9sf2', 'C4bpa', 'Ank3', 'Afap1', 'Cmas', 'Tmpo', 'Lzic', 'Lgals1', 'Eps8l2', 'Krt7', 'Aif1l', 'Maged2', 'Fbln1', 'Pitrm1', 'Ctsm', 'Lta4h', 'Septin5', 'Uba6', 'Slc4a2', 'Nln', 'Uggt1', 'Chdh', 'Ceacam11', 'Loxl2', 'Fam136a', 'Cd5l', 'Dbnl', 'Sec11a', 'AABR07060872.1', 'Casp6', 'Lima1', 'Emb', 'Lgals5', 'Phactr4', 'Psg19', 'Matr3', 'Hpx', 'Vcpip1', 'Serpinb9', 'Arsb', 'Lrrc59', 'Hdhd2', 'Prl8a9', 'Kng1', 'Prl4a1', 'Nars1', 'Aqp1', 'Gab1', 'Myo18a', 'Prl6a1', 'Klhl22', 'Tnfaip8', 'Agt', 'Abcb4', 'Slc2a1', 'Plcd1', 'Tes', 'Creg1', 'Naxd', 'Orm1', 'LOC299282', 'Fndc3a', 'Slc27a3', 'Retsat', 'Pdlim4', 'Gnb4', 'St6gal1', 'Prss8', 'Arpp19', 'Bgn', 'Crtap', 'Prrc1', 'Cbr1l1', 'Lamp1', 'Fblim1', 'Abracl', 'Dpp3', 'Vnn1', 'Grn', 'Ddah1', 'Inf2', 'Timp3', 'Xdh', 'Sgpl1', 'Sh3bgrl', 'Etfdh', 'Sars1', 'Cobll1', 'Pgk1', 'Man2a1', 'Nostrin', 'Tspan9', 'Limd1', 'Lad1', 'Mesd', 'Gar1', 'Pon2', 'Snx12', 'Slc38a4', 'Cand2', 'Plvap', 'Tbl2', 'Atp2b1', 'Pdxk', 'Creld2', 'Hp', 'Ctsr', 'Pcmt1', 'RGD1562844', 'Ccdc51', 'Cotl1', 'Sbsn', 'Cpne1', 'Cnn3', 'Psgb1', 'Tram2', 'Lipa', 'Hsd17b11', 'Fnta', 'Ppl', 'Ctla2a', 'Lrpap1', 'Hmgb3', 'Slc40a1', 'Scamp1', 'Rbp1', 'Igfbp2', 'Erp44', 'Lgalsl', 'Ptpn23', 'Stmn1', 'Eml2', 'Fam122b', 'Sparc', 'Abhd12', 'Pecam', 'C4a', 'Prxl2a', 'Prkar2a', 'Csrp2', 'Glg1', 'Ifnk', 'Colec12', 'Cox5b', 'Ctsql2', 'Prl3d4', 'Camk2d', 'Npc2', 'Itih1', 'Pltp', 'Ckap4', 'Syap1', 'Ngp', 'Prl8a2', 'Epb41l3', 'Cubn', 'Pappa2', 'Mustn1', 'Adh1', 'LOC102557244'}\n"
     ]
    }
   ],
   "source": [
    "## Pre-eclampsia\n",
    "display(preeclam_ppi_data.head(10))\n",
    "\n",
    "preeclam_genes_in_ppi = set(preeclam_ppi_data['node1'].values.tolist() + preeclam_ppi_data['node2'].values.tolist())\n",
    "preeclam_all_genes = set(preeclam_expression_data['gene'].values.tolist())\n",
    "print(f'Pre-eclampsia -> # Total genes = {len(preeclam_all_genes)}, # Genes in PPI = {len(preeclam_genes_in_ppi)}')\n",
    "\n",
    "print(f'Genes NOT in ppi ({len(preeclam_all_genes - preeclam_genes_in_ppi)}):')\n",
    "print(preeclam_all_genes - preeclam_genes_in_ppi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pre-eclampsia -> # of Edges = 315, # of Nodes = 153\n"
     ]
    }
   ],
   "source": [
    "unique_nodes = set(preeclam_ppi_data['node1'].values.tolist() + preeclam_ppi_data['node2'].values.tolist())\n",
    "print(f'Pre-eclampsia -> # of Edges = {len(preeclam_ppi_data)}, # of Nodes = {len(unique_nodes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Verificar como podem existir Genes no STRING que no esto na tabela de expresso diferencial\n",
    "# preeclam_ppi_data = preeclam_ppi_data[preeclam_ppi_data['node1'].isin(preeclam_all_genes) & preeclam_ppi_data['node2'].isin(preeclam_all_genes)]\n",
    "\n",
    "\n",
    "\n",
    "# preeclam_genes_in_ppi = set(preeclam_ppi_data['node1'].values.tolist() + preeclam_ppi_data['node2'].values.tolist())\n",
    "# preeclam_all_genes = set(preeclam_expression_data['gene'].values.tolist())\n",
    "# print(f'Cancer -> # Total genes = {len(preeclam_all_genes)}, # Genes in PPI = {len(preeclam_genes_in_ppi)}')\n",
    "\n",
    "# print(f'Genes NOT in ppi ({len(preeclam_all_genes - preeclam_genes_in_ppi)}):')\n",
    "# print(preeclam_all_genes - preeclam_genes_in_ppi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>Actg2</td>\n",
       "      <td>10116.ENSRNOP00000024084</td>\n",
       "      <td>10116.ENSRNOP00000050322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Actc1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acta2</td>\n",
       "      <td>Tln1</td>\n",
       "      <td>10116.ENSRNOP00000073101</td>\n",
       "      <td>10116.ENSRNOP00000022401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>Stk10</td>\n",
       "      <td>Tfrc</td>\n",
       "      <td>10116.ENSRNOP00000044325</td>\n",
       "      <td>10116.ENSRNOP00000002407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>Tmed3</td>\n",
       "      <td>Uso1</td>\n",
       "      <td>10116.ENSRNOP00000018603</td>\n",
       "      <td>10116.ENSRNOP00000003277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>Tmed3</td>\n",
       "      <td>Tmed7</td>\n",
       "      <td>10116.ENSRNOP00000018603</td>\n",
       "      <td>10116.ENSRNOP00000004942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>Tmed7</td>\n",
       "      <td>Uso1</td>\n",
       "      <td>10116.ENSRNOP00000004942</td>\n",
       "      <td>10116.ENSRNOP00000003277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>Tuba4a</td>\n",
       "      <td>Tubb3</td>\n",
       "      <td>10116.ENSRNOP00000004797</td>\n",
       "      <td>10116.ENSRNOP00000023452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      node1  node2                  node1_id                  node2_id\n",
       "0     Acta1  Actc1  10116.ENSRNOP00000024084  10116.ENSRNOP00000011773\n",
       "1     Acta1  Acta2  10116.ENSRNOP00000024084  10116.ENSRNOP00000073101\n",
       "2     Acta1  Actg2  10116.ENSRNOP00000024084  10116.ENSRNOP00000050322\n",
       "3     Acta2  Actc1  10116.ENSRNOP00000073101  10116.ENSRNOP00000011773\n",
       "4     Acta2   Tln1  10116.ENSRNOP00000073101  10116.ENSRNOP00000022401\n",
       "..      ...    ...                       ...                       ...\n",
       "495   Stk10   Tfrc  10116.ENSRNOP00000044325  10116.ENSRNOP00000002407\n",
       "496   Tmed3   Uso1  10116.ENSRNOP00000018603  10116.ENSRNOP00000003277\n",
       "497   Tmed3  Tmed7  10116.ENSRNOP00000018603  10116.ENSRNOP00000004942\n",
       "498   Tmed7   Uso1  10116.ENSRNOP00000004942  10116.ENSRNOP00000003277\n",
       "499  Tuba4a  Tubb3  10116.ENSRNOP00000004797  10116.ENSRNOP00000023452\n",
       "\n",
       "[500 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#Duplicated Edges:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "node1    node2 \n",
       "Dnajb11  Ruvbl2    2\n",
       "Ehd1     Ehd4      2\n",
       "Tuba4a   Tubb3     2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge STRING networks and check/remove for duplicates\n",
    "merged_ppi_data = pd.concat([cancer_ppi_data, preeclam_ppi_data], axis=0, join='outer', ignore_index=True)\n",
    "display(merged_ppi_data)\n",
    "print('#Duplicated Edges:')\n",
    "merged_ppi_data.pivot_table(index = ['node1', 'node2'], aggfunc ='size')[merged_ppi_data.pivot_table(index = ['node1', 'node2'], aggfunc ='size') > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of Edges = 497, # of Nodes = 243\n"
     ]
    }
   ],
   "source": [
    "merged_ppi_data.drop_duplicates(keep='first', ignore_index=True, inplace=True)\n",
    "unique_nodes = set(merged_ppi_data['node1'].values.tolist() + merged_ppi_data['node2'].values.tolist())\n",
    "print(f'# of Edges = {len(merged_ppi_data)}, # of Nodes = {len(unique_nodes)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "ppi_fn = 'ppi_data'\n",
    "cancer_ppi_data.to_csv(Path(output_path, f'./{ppi_fn}_cancer.csv'), index=False)\n",
    "preeclam_ppi_data.to_csv(Path(output_path, f'./{ppi_fn}_pre_eclampsia.csv'), index=False)\n",
    "merged_ppi_data.to_csv(Path(output_path, f'./{ppi_fn}_merged.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load and process KEGG enrichment (pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_rename_cols = {\n",
    "    '#term ID' : 'pathway_id',\n",
    "    'term description' : 'pathway_name',\n",
    "    'observed gene count' : 'observed_gene_count',\n",
    "    'background gene count' : 'background_gene_count',\n",
    "    'strength' : 'strength',\n",
    "    'false discovery rate' : 'fdr',\n",
    "    'matching proteins in your network (labels)' : 'gene_labels_in'\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04540</td>\n",
       "      <td>Gap junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04810</td>\n",
       "      <td>Regulation of actin cytoskeleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04530</td>\n",
       "      <td>Tight junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04210</td>\n",
       "      <td>Apoptosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rno04270</td>\n",
       "      <td>Vascular smooth muscle contraction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rno04611</td>\n",
       "      <td>Platelet activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rno04961</td>\n",
       "      <td>Endocrine and other factor-regulated calcium r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rno00020</td>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rno00020</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rno03010</td>\n",
       "      <td>Ribosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rno04510</td>\n",
       "      <td>Focal adhesion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rno04144</td>\n",
       "      <td>Endocytosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pathway_id                                       pathway_name\n",
       "0    rno04145                                          Phagosome\n",
       "1    rno04540                                       Gap junction\n",
       "2    rno04810                   Regulation of actin cytoskeleton\n",
       "3    rno04530                                     Tight junction\n",
       "4    rno04210                                          Apoptosis\n",
       "5    rno04270                 Vascular smooth muscle contraction\n",
       "6    rno04141        Protein processing in endoplasmic reticulum\n",
       "7    rno04611                                Platelet activation\n",
       "8    rno04961  Endocrine and other factor-regulated calcium r...\n",
       "9    rno00020                          Citrate cycle (TCA cycle)\n",
       "10   rno00020                                  Carbon metabolism\n",
       "11   rno03010                                           Ribosome\n",
       "12   rno04510                                     Focal adhesion\n",
       "13   rno04144                                        Endocytosis\n",
       "14   rno01230                        Biosynthesis of amino acids\n",
       "15   rno00010                       Glycolysis / Gluconeogenesis"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Cancer\n",
    "cancer_pathway_data = pd.read_csv(kegg_cancer_path, sep=';', usecols=list(pathway_rename_cols.keys())).rename(columns=pathway_rename_cols)\n",
    "cancer_pathway_data['pathway_name'] = cancer_pathway_data['pathway_name'].str.replace(',', ';')\n",
    "cancer_pathway_data[['pathway_id', 'pathway_name']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tuba4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Ctsl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Cs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Pfkl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Aldh2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>LOC100911625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Ldhb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                  pathway_name gene_labels_in\n",
       "0     rno04145                     Phagosome           Tfrc\n",
       "1     rno04145                     Phagosome         Tuba4a\n",
       "2     rno04145                     Phagosome          Tubb3\n",
       "3     rno04145                     Phagosome          Tubb6\n",
       "4     rno04145                     Phagosome           Ctsl\n",
       "..         ...                           ...            ...\n",
       "116   rno01230   Biosynthesis of amino acids             Cs\n",
       "117   rno00010  Glycolysis / Gluconeogenesis           Pfkl\n",
       "118   rno00010  Glycolysis / Gluconeogenesis          Aldh2\n",
       "119   rno00010  Glycolysis / Gluconeogenesis   LOC100911625\n",
       "120   rno00010  Glycolysis / Gluconeogenesis           Ldhb\n",
       "\n",
       "[121 rows x 3 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_pathway_net = cancer_pathway_data[['pathway_id', 'pathway_name', 'gene_labels_in']].copy()\n",
    "cancer_pathway_net['gene_labels_in'] = cancer_pathway_net['gene_labels_in'].str.split(',')\n",
    "cancer_pathway_net = cancer_pathway_net.explode(column='gene_labels_in', ignore_index=True)\n",
    "cancer_pathway_net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno01100</td>\n",
       "      <td>Metabolic pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno03010</td>\n",
       "      <td>Ribosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04066</td>\n",
       "      <td>HIF-1 signaling pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rno03060</td>\n",
       "      <td>Protein export</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rno01200</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rno04142</td>\n",
       "      <td>Lysosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rno04922</td>\n",
       "      <td>Glucagon signaling pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pathway_id                                 pathway_name\n",
       "0    rno04141  Protein processing in endoplasmic reticulum\n",
       "1    rno01100                           Metabolic pathways\n",
       "2    rno00010                 Glycolysis / Gluconeogenesis\n",
       "3    rno03010                                     Ribosome\n",
       "4    rno04066                      HIF-1 signaling pathway\n",
       "5    rno03060                               Protein export\n",
       "6    rno01230                  Biosynthesis of amino acids\n",
       "7    rno01200                            Carbon metabolism\n",
       "8    rno04145                                    Phagosome\n",
       "9    rno04142                                     Lysosome\n",
       "10   rno04922                   Glucagon signaling pathway\n",
       "11   rno04216                                  Ferroptosis"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Pre-eclampsia\n",
    "preeclam_pathway_data = pd.read_excel(kegg_preeclam_path, usecols=list(pathway_rename_cols.keys())).rename(columns=pathway_rename_cols)\n",
    "preeclam_pathway_data['pathway_name'] = preeclam_pathway_data['pathway_name'].str.replace(',', ';')\n",
    "preeclam_pathway_data[['pathway_id', 'pathway_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Hsph1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Erp29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Dnajb11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Sec31a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Calr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>rno04922</td>\n",
       "      <td>Glucagon signaling pathway</td>\n",
       "      <td>Slc2a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc40a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Hmox1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc3a2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                                 pathway_name gene_labels_in\n",
       "0     rno04141  Protein processing in endoplasmic reticulum          Hsph1\n",
       "1     rno04141  Protein processing in endoplasmic reticulum          Erp29\n",
       "2     rno04141  Protein processing in endoplasmic reticulum        Dnajb11\n",
       "3     rno04141  Protein processing in endoplasmic reticulum         Sec31a\n",
       "4     rno04141  Protein processing in endoplasmic reticulum           Calr\n",
       "..         ...                                          ...            ...\n",
       "140   rno04922                   Glucagon signaling pathway         Slc2a1\n",
       "141   rno04216                                  Ferroptosis           Tfrc\n",
       "142   rno04216                                  Ferroptosis        Slc40a1\n",
       "143   rno04216                                  Ferroptosis          Hmox1\n",
       "144   rno04216                                  Ferroptosis         Slc3a2\n",
       "\n",
       "[145 rows x 3 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preeclam_pathway_net = preeclam_pathway_data[['pathway_id', 'pathway_name', 'gene_labels_in']].copy()\n",
    "preeclam_pathway_net['gene_labels_in'] = preeclam_pathway_net['gene_labels_in'].str.split(',')\n",
    "preeclam_pathway_net = preeclam_pathway_net.explode(column='gene_labels_in', ignore_index=True)\n",
    "preeclam_pathway_net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cancer -> # of pathways = 16, # of genes = 67, # of edges = 121\n",
      "Pre-eclampsia -> # of pathways = 12, # of genes = 103, # of edges = 145\n"
     ]
    }
   ],
   "source": [
    "print(f\"Cancer -> # of pathways = {len(cancer_pathway_net['pathway_name'].unique())}, \"\n",
    "      + f\"# of genes = {len(cancer_pathway_net['gene_labels_in'].unique())}, # of edges = {len(cancer_pathway_net)}\")\n",
    "\n",
    "print(f\"Pre-eclampsia -> # of pathways = {len(preeclam_pathway_net['pathway_name'].unique())}, \"\n",
    "      + f\"# of genes = {len(preeclam_pathway_net['gene_labels_in'].unique())}, # of edges = {len(preeclam_pathway_net)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tuba4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tubb6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Ctsl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>rno04922</td>\n",
       "      <td>Glucagon signaling pathway</td>\n",
       "      <td>Slc2a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Tfrc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc40a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Hmox1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>rno04216</td>\n",
       "      <td>Ferroptosis</td>\n",
       "      <td>Slc3a2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>266 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                pathway_name gene_labels_in\n",
       "0     rno04145                   Phagosome           Tfrc\n",
       "1     rno04145                   Phagosome         Tuba4a\n",
       "2     rno04145                   Phagosome          Tubb3\n",
       "3     rno04145                   Phagosome          Tubb6\n",
       "4     rno04145                   Phagosome           Ctsl\n",
       "..         ...                         ...            ...\n",
       "261   rno04922  Glucagon signaling pathway         Slc2a1\n",
       "262   rno04216                 Ferroptosis           Tfrc\n",
       "263   rno04216                 Ferroptosis        Slc40a1\n",
       "264   rno04216                 Ferroptosis          Hmox1\n",
       "265   rno04216                 Ferroptosis         Slc3a2\n",
       "\n",
       "[266 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "pathway_name                                 gene_labels_in\n",
       "Glycolysis / Gluconeogenesis                 Aldh2             2\n",
       "                                             Ldhb              2\n",
       "Phagosome                                    Tfrc              2\n",
       "                                             Tuba4a            2\n",
       "                                             Tubb3             2\n",
       "Protein processing in endoplasmic reticulum  Dnajb11           2\n",
       "                                             Ero1a             2\n",
       "                                             Lman2             2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge both pathway nets and check/remove for duplicates\n",
    "\n",
    "merged_pathway_net = pd.concat([cancer_pathway_net, preeclam_pathway_net], axis=0, join='outer', ignore_index=True)\n",
    "display(merged_pathway_net)\n",
    "print()\n",
    "merged_pathway_net.pivot_table(index = ['pathway_name', 'gene_labels_in'], aggfunc ='size')[merged_pathway_net.pivot_table(index = ['pathway_name', 'gene_labels_in'], aggfunc ='size') > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_pathway_net.drop_duplicates(keep='first', ignore_index=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged net -> # of pathways = 22, # of genes = 162, # of edges = 258\n"
     ]
    }
   ],
   "source": [
    "print(f\"Merged net -> # of pathways = {len(merged_pathway_net['pathway_name'].unique())}, \"\n",
    "      + f\"# of genes = {len(merged_pathway_net['gene_labels_in'].unique())}, # of edges = {len(merged_pathway_net)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_net_fn = 'kegg_net'\n",
    "cancer_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_cancer.csv'), index=False)\n",
    "preeclam_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_pre_eclampsia.csv'), index=False)\n",
    "merged_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_merged.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### KEGG compartilhadas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comp_pathway_data = pd.read_excel(kegg_common_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Hsph1,Erp29,Dnajb11,Sec31a,Calr,Pdia4,Ero1a,Ck...</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Dnajb11,Ero1a,Nsfl1c,Lman2,Ubqln4,P4hb,Ddost,R...</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Shmt2,Idh1,Got1,Phgdh,Aldoc,Pgk1</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno01230</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>Pfkl,Rpia,Mat2a,Cs</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno01200</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "      <td>Hk2,Shmt2,Idh1,Got1,Phgdh,Aldoc,Pgk1</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rno01200</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "      <td>Pfkl,Rpia,Sdha,Cs,Suclg2,Ogdh</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rno03010</td>\n",
       "      <td>Ribosome</td>\n",
       "      <td>Rpl24,Rpl32,Rpl22,Rpl23a,Rps13,Rpl30,Rpl14</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rno03010</td>\n",
       "      <td>Ribosome</td>\n",
       "      <td>Rpl19,Rpl34,Rpl35,Rpl13,Rpl18a,Rps25,Rps15,Rpl...</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tfrc,Tuba4a,Tubb3,Tubb6,Ctsl,Tuba3b,Tuba8,C3,T...</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rno04145</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>Tfrc,Calr,Tuba4a,Sec61a1,Tubb3,Lamp1,RT1-CE1,C...</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Pfkl,Aldh2,LOC100911625,Ldhb</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Aldh2,Aldh9a1,Hk2,Pgm1,Ldhb,Aldoc,Pgk1</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pathway_id                                 pathway_name  \\\n",
       "0    rno04141  Protein processing in endoplasmic reticulum   \n",
       "1    rno04141  Protein processing in endoplasmic reticulum   \n",
       "2    rno01230                  Biosynthesis of amino acids   \n",
       "3    rno01230                  Biosynthesis of amino acids   \n",
       "4    rno01200                            Carbon metabolism   \n",
       "5    rno01200                            Carbon metabolism   \n",
       "6    rno03010                                     Ribosome   \n",
       "7    rno03010                                     Ribosome   \n",
       "8    rno04145                                    Phagosome   \n",
       "9    rno04145                                    Phagosome   \n",
       "10   rno00010                 Glycolysis / Gluconeogenesis   \n",
       "11   rno00010                 Glycolysis / Gluconeogenesis   \n",
       "\n",
       "                                       gene_labels_in        disease  \n",
       "0   Hsph1,Erp29,Dnajb11,Sec31a,Calr,Pdia4,Ero1a,Ck...  pre_eclampsia  \n",
       "1   Dnajb11,Ero1a,Nsfl1c,Lman2,Ubqln4,P4hb,Ddost,R...         Cancer  \n",
       "2                    Shmt2,Idh1,Got1,Phgdh,Aldoc,Pgk1  pre_eclampsia  \n",
       "3                                  Pfkl,Rpia,Mat2a,Cs         Cancer  \n",
       "4                Hk2,Shmt2,Idh1,Got1,Phgdh,Aldoc,Pgk1  pre_eclampsia  \n",
       "5                       Pfkl,Rpia,Sdha,Cs,Suclg2,Ogdh         Cancer  \n",
       "6          Rpl24,Rpl32,Rpl22,Rpl23a,Rps13,Rpl30,Rpl14         Cancer  \n",
       "7   Rpl19,Rpl34,Rpl35,Rpl13,Rpl18a,Rps25,Rps15,Rpl...  pre_eclampsia  \n",
       "8   Tfrc,Tuba4a,Tubb3,Tubb6,Ctsl,Tuba3b,Tuba8,C3,T...         Cancer  \n",
       "9   Tfrc,Calr,Tuba4a,Sec61a1,Tubb3,Lamp1,RT1-CE1,C...  pre_eclampsia  \n",
       "10                       Pfkl,Aldh2,LOC100911625,Ldhb         Cancer  \n",
       "11             Aldh2,Aldh9a1,Hk2,Pgm1,Ldhb,Aldoc,Pgk1  pre_eclampsia  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KEGG compartilhadas\n",
    "pathway_rename_cols = {\n",
    "    '#term ID' : 'pathway_id',\n",
    "    'term description' : 'pathway_name',\n",
    "    'observed gene count' : 'observed_gene_count',\n",
    "    'background gene count' : 'background_gene_count',\n",
    "    'strength' : 'strength',\n",
    "    'false discovery rate' : 'fdr',\n",
    "    'matching proteins in your network (labels)' : 'gene_labels_in',\n",
    "    'Unnamed: 8' : 'disease'\n",
    "}\n",
    "\n",
    "comp_pathway_data = pd.read_excel(kegg_common_path, usecols=list(pathway_rename_cols.keys())).rename(columns=pathway_rename_cols)\n",
    "comp_pathway_data['pathway_name'] = comp_pathway_data['pathway_name'].str.replace(',', ';')\n",
    "comp_pathway_data[['pathway_id', 'pathway_name', 'gene_labels_in', 'disease']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_labels_in</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Hsph1</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Erp29</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Dnajb11</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Sec31a</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rno04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Calr</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Hk2</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Pgm1</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Ldhb</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Aldoc</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>rno00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Pgk1</td>\n",
       "      <td>pre_eclampsia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                                 pathway_name gene_labels_in  \\\n",
       "0     rno04141  Protein processing in endoplasmic reticulum          Hsph1   \n",
       "1     rno04141  Protein processing in endoplasmic reticulum          Erp29   \n",
       "2     rno04141  Protein processing in endoplasmic reticulum        Dnajb11   \n",
       "3     rno04141  Protein processing in endoplasmic reticulum         Sec31a   \n",
       "4     rno04141  Protein processing in endoplasmic reticulum           Calr   \n",
       "..         ...                                          ...            ...   \n",
       "97    rno00010                 Glycolysis / Gluconeogenesis            Hk2   \n",
       "98    rno00010                 Glycolysis / Gluconeogenesis           Pgm1   \n",
       "99    rno00010                 Glycolysis / Gluconeogenesis           Ldhb   \n",
       "100   rno00010                 Glycolysis / Gluconeogenesis          Aldoc   \n",
       "101   rno00010                 Glycolysis / Gluconeogenesis           Pgk1   \n",
       "\n",
       "           disease  \n",
       "0    pre_eclampsia  \n",
       "1    pre_eclampsia  \n",
       "2    pre_eclampsia  \n",
       "3    pre_eclampsia  \n",
       "4    pre_eclampsia  \n",
       "..             ...  \n",
       "97   pre_eclampsia  \n",
       "98   pre_eclampsia  \n",
       "99   pre_eclampsia  \n",
       "100  pre_eclampsia  \n",
       "101  pre_eclampsia  \n",
       "\n",
       "[102 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_pathway_net = comp_pathway_data[['pathway_id', 'pathway_name', 'gene_labels_in', 'disease']].copy()\n",
    "comp_pathway_net['gene_labels_in'] = comp_pathway_net['gene_labels_in'].str.split(',')\n",
    "comp_pathway_net = comp_pathway_net.explode(column='gene_labels_in', ignore_index=True)\n",
    "comp_pathway_net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared -> # of pathways = 6, # of genes = 79, # of edges = 102\n"
     ]
    }
   ],
   "source": [
    "print(f\"Shared -> # of pathways = {len(comp_pathway_net['pathway_name'].unique())}, \"\n",
    "      + f\"# of genes = {len(comp_pathway_net['gene_labels_in'].unique())}, # of edges = {len(comp_pathway_net)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared Cancer -> # of pathways = 6, # of genes = 38, # of edges = 42\n",
      "Shared Pre-eclampsia -> # of pathways = 6, # of genes = 49, # of edges = 60\n"
     ]
    }
   ],
   "source": [
    "comp_pathway_net_cancer = comp_pathway_net[comp_pathway_net['disease'] == 'Cancer']\n",
    "comp_pathway_net_preeclam = comp_pathway_net[comp_pathway_net['disease'] == 'pre_eclampsia']\n",
    "\n",
    "print(f\"Shared Cancer -> # of pathways = {len(comp_pathway_net_cancer['pathway_name'].unique())}, \"\n",
    "      + f\"# of genes = {len(comp_pathway_net_cancer['gene_labels_in'].unique())}, # of edges = {len(comp_pathway_net_cancer)}\")\n",
    "\n",
    "print(f\"Shared Pre-eclampsia -> # of pathways = {len(comp_pathway_net_preeclam['pathway_name'].unique())}, \"\n",
    "      + f\"# of genes = {len(comp_pathway_net_preeclam['gene_labels_in'].unique())}, # of edges = {len(comp_pathway_net_preeclam)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_net_fn = 'kegg_net'\n",
    "# comp_pathway_net_cancer.to_csv(Path(output_path, f'./{pathway_net_fn}_compartilhadas_cancer.csv'), index=False)\n",
    "# comp_pathway_net_preeclam.to_csv(Path(output_path, f'./{pathway_net_fn}_compartilhadas_preeclam.csv'), index=False)\n",
    "\n",
    "comp_pathway_net.to_csv(Path(output_path, f'./{pathway_net_fn}_compartilhadas.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
